### DISCUSSION TITLE
Introduce some gRPC?

### DISCUSSION THREAD

**Discussion by edeandrea** [quarkusio/quarkus-super-heroes#379] [Ideas]
Should we introduce some gRPC into some of the services? Maybe a matching gRPC layer in the heroes/villains? We're already showing rest in `rest-fights` & `rest-narration`, so maybe for east/west traffic between `rest-fights` and `rest-heroes`/`rest-villains` we go to gRPC?

@cescoffier / @agoncal / @holly-cummins what do you all think?

**Comment by edeandrea** [comment]
Although by using rest in heroes/villains, in `rest-fights` we're showing how to use the rest client and I'm not sure we'd want to get rid of that.

Maybe in heroes & villains we have both rest & gRPC, then maybe in `rest-fights` a runtime flag which tells which client to use (rest vs gRPC)?

@holly-cummins This would also bring in the ability to use the [Pact gRPC plugin](https://docs.pact.io/implementation_guides/pact_plugins/examples/grpc)

**Comment by agoncal** [comment]
What about introducing a new microservice ? 

With @cescoffier we once had this idea of having another microservice that would return a "location/universe/planet" where the fight happens. That would return an image (the image of the planet, location....that would be displayed in the UI as a background) and a name/description of the location ("Abafar is a desert planet located in the Outer Rim with a completely white surface..."). We would have a new button that would pick a random location. This way you can pick a random villain, a random hero, and make them fight on a random location.

WDYT ? And having gRPC could be something we could also add to the workshop

**Reply by cescoffier** [reply]
Another idea I had recently, is to have a variant of the reactive crud service using gRPC instead of the reactive rest approach. The rest of the code stay unchanged.

**Reply by edeandrea** [reply]
I think we should not have different variants of the different services. I think that is too much of a maintenance headache, especially since we produce & maintain container images. We'd have to maintain variants of the container images too.

I think there is value in showing both blocking & reactive rest endpoints as well as both reactive & blocking rest clients (like we have today). It's also easy to run benchmarks and show performance differences between fully blocking & fully reactive, since both villains & heroes use the same schema with the same amount of data. They are essentially the same service.

**Reply by edeandrea** [reply]
@agoncal / @cescoffier thinking down the location service rabbit hole a bit, do you have in mind a set of locations & images to pick from? We'd probably need quite a few.

**Comment by cescoffier** [comment]
No, I don’t have any pictures. I was thinking taking pictures of famous
cities and places in the world, plus a few places related to super heroes
(Metropolis, Gotham City, …).

We do not need a lot. Something like 50 would be enough.

**Reply by edeandrea** [reply]
Yeah 50 was the number I had in mind. The site where we scraped all the data from to begin with - do they have a "universe" of locations?

**Reply by edeandrea** [reply]
I found this....

https://www.superherodb.com/locations/

**Reply by cescoffier** [reply]
Awesome! We can take them from there!

**Reply by edeandrea** [reply]
Although after looking at those it doesn't seem very useful. There are 1600+ of them listed, many aren't associated with any characters nor do they have any pictures. And even those that are, I'm not sure how relevant the images are.

See https://www.superherodb.com/gotham-city/700-378/gallery/ for example.

**Reply by edeandrea** [reply]
If we're going to go down this road of a location service, should we build it in Kotlin? Kill 2 birds with one stone?

**Reply by agoncal** [reply]
Or just images of space, such as https://esahubble.org/images/archive/top100/

**Reply by agoncal** [reply]
Kotlin ? Well, I suppose it's quite useful to show the Kotlin support in Quarkus

**Reply by edeandrea** [reply]
I think having a service written in Kotlin would be helpful. I've also been looking for a way to integrate Microcks into the superheroes, so having an additional service would be helpful there too.

**Reply by edeandrea** [reply]
Here's an initial commit I worked on today. Not done yet but close - https://github.com/edeandrea/quarkus-super-heroes/tree/location-service/location-service

Kotlin & gRPC

**Reply by agoncal** [reply]
In the other services we've added the technology in their name (eg. `rest-heroes` or `event-statistics`, etc), what about calling it `grpc-location` ? 

Oh, that brings me to the Narration. Maybe we should call it `ai-narration`. WDYT ?

**Reply by edeandrea** [reply]
Yeah I was thinking about that last night and was planning to rename it today.

As for narration, it is a restful service so I think rest-narration is appropriate? We don't name the other services based on the technologies they use (like mongodb-fights).

**Reply by cescoffier** [reply]
I like the idea of introducing gRPC and Kotlin. However, the combination can be slightly weird as we do not support kotlin-grpc. So, kotlin users might be slightly confused.

**Reply by edeandrea** [reply]
I guess I'm not sure what kotlin-grpc is. Is that some other non-quarkus library?

**Reply by cescoffier** [reply]
Yes, it’s not supported in Quarkus. It’s the traditional way to use gRPC in Kotlin.

**Reply by edeandrea** [reply]
Looking at https://grpc.io/docs/languages/kotlin/quickstart/ it doesn't look a whole lot different than what I'm doing...(https://github.com/edeandrea/quarkus-super-heroes/blob/location-service/grpc-locations/src/main/kotlin/io/quarkus/sample/superheroes/location/grpc/LocationGrpcService.kt for the server & https://github.com/edeandrea/quarkus-super-heroes/blob/location-service/grpc-locations/src/test/kotlin/io/quarkus/sample/superheroes/location/grpc/LocationGrpcServiceTests.kt for the client)

**Reply by edeandrea** [reply]
@cescoffier / @agoncal would you be able to help me get some pictures/descriptions of locations? I've started...

https://github.com/quarkusio/quarkus-super-heroes/blob/characterdata/all-locations.sql & https://github.com/quarkusio/quarkus-super-heroes/tree/characterdata/images/locations

You can either commit directly to the `characterdata` branch, or do a PR and I can merge.

**Comment by edeandrea** [comment]
I've created this as an issue now that its underway.
